File: go/2022/day01/day01.go
package main

import (
	"aoc/cast"
	"aoc/mathy"
	"aoc/utils"
	"flag"
	"fmt"
	"os"
	"sort"
	"strings"
)

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)
	if err != nil {
		panic("Unable to read the input file")
	}

	input := string(bytes)

	fmt.Printf("AoC 2022, Day1, Part1 solution is: %d \n", part1(input))
	fmt.Printf("AoC 2022, Day1, Part2 solution is: %d \n", part2(input))
}

func part1(input string) int {
	return utils.MaxInt(parseInput(input)...)
}

func part2(input string) int {
	calories := parseInput(input)
	sort.Ints(sort.IntSlice(calories))

	total := 0
	for i := len(calories) - 1; i >= len(calories)-3; i-- {
		total += calories[i]
	}
	return total
}

func parseInput(input string) []int {
	blocks := strings.Split(input, "\n\n")

	items := []int{}

	for _, block := range blocks {
		lines := strings.Split(block, "\n")
		calories := cast.MapSliceOfStringToInt(lines)

		items = append(items, mathy.SumIntSlice(calories))
	}

	return items
}

File: python/2022/day01/day01.py
from typing import List


def part1(input) -> int:
    loads = get_calories(input)
    return max(loads)


def part2(input: str) -> int:
    loads = get_calories(input)
    loads.sort()
    return sum(loads[-3:])


def parse_input(input: str) -> List[str]:
    with open(input) as fn:
        raw_data = fn.read().strip()
        parts = raw_data.split("\n\n")
    return parts


def get_calories(input: str) -> List[int]:
    loads = []
    for part in parse_input(input):
        calories = [
            int(x) for x in part.split()
        ]  # this can be written as: map(int, part.split())
        loads.append(sum(calories))
    return loads


def main():
    puzzle_input_path = "2022/day01/input.txt"
    print(f"AoC2022 Day1, Part1 solution is: {part1(puzzle_input_path)}")
    print(f"AoC2022 Day1, Part2 solution is: {part2(puzzle_input_path)}")


if __name__ == "__main__":
    main()
